scenes/cslib_object.txt

```diff
@@ -0,0 +1,111 @@
+*comment cslib_object
+*comment Allows the simple manipulation of arrays of sets of variables
+*comment @StephenHart
+*comment
+*comment How it works:
+*comment A object is always accessed through its main instance.
+*comment When you "get" an instance of a object, the data for that particular instance 
+*comment is copied into the main instance
+*comment
+*comment A object definition has the following elements defined in startup.txt:
+*comment	1. A object "fieldname" array defining all the field names
+*comment		objectname_fieldname_N fieldname
+*comment	2. A main instance of this object. This is used to 'return'
+*comment		the object (similar to cslib_ret). Also includes the index
+*comment		of the current instance
+*comment	3. An array of instances of the object
+*comment		
+*comment Public methods
+*comment	get object index
+*comment		copy the instance of the object at index into main instance
+*comment	set object index field value
+*comment		set the specified field to the specified value in the instance
+*comment		of the object at index
+*comment
+*comment ------------------------------------------------------------------
+*comment EXAMPLE - startup.txt - object "dragon"
+*comment 
+*comment *create dragon_fieldname_1 	"name"
+*comment *create dragon_fieldname_2 	"hp"
+*comment *create dragon_fieldname_max	2
+*comment
+*comment *create dragon_name		""
+*comment *create dragon_hp			""
+*comment
+*comment *create dragon_1_name 		"Wally"
+*comment *create dragon_1_hp 		100
+*comment *create dragon_2_name 		"Dave"
+*comment *create dragon_2_hp		150
+*comment
+*comment EXAMPLE of getting a dragon instance
+*comment
+*comment *gosub_scene cslib_object get "dragon" 2
+*comment 	sets dragon_name to "Dave" and dragon_hp to 150 and dragon_index to 2
+*comment
+*comment EXAMPLE of setting a dragon instance
+*comment
+*comment *gosub_scene cslib_object set "dragon" 2 "name" "Sir David"
+*comment    sets dragon_2_name to "Sir David"
+*comment	now when you get dragon 2, dragon_name will be "Sir David"
+*comment	this will persist until you set it again
+
+*comment GET
+*comment ------------------------------------------------------------------
+*comment Takes the instance of a object specified by the index and copies
+*comment the data into the main instance
+*comment ------------------------------------------------------------------
+*comment
+*comment params:
+*comment	p_objectname (string): the name of the object
+*comment	p_index (integer): index of this object instance
+*comment returns:
+*comment	none
+*comment usage example:
+*comment 	cslib_object get "dragon" 2
+*comment ------------------------------------------------------------------
+
+*label get
+*params p_objectname p_index
+*set cslib_ret ""
+*gosub _set_main_instance p_objectname p_index
```

Comment:
I'm tempted to say we should guard here against bad IDs, but that would force users into using array semantics (i.e. their object definitions would have to have a _max and _count variable), like:

`
*comment *create dragon_fieldname_1 	"name"
 *comment *create dragon_fieldname_2 	"hp"
 *comment *create dragon_fieldname_max	2
 *comment
 *comment *create dragon_name		""
 *comment *create dragon_hp			""
 *comment
 *comment *create dragon_1_name 		"Wally"
 *comment *create dragon_1_hp 		100
 *comment *create dragon_2_name 		"Dave"
 *comment *create dragon_2_hp		150
 *comment *create_dragon_count              2
 *comment *create_dragon_max                 5                        (there's room for 3 more to be generated at runtime)
 `

Reply:
I had a look at cslib_array and it is pretty impressive. I thought about trying to integrate it but decided against it.
My thoughts with cslib_object was to keep it as simple as possible. The target user is a non-programmer.
The only 'new' thing for them is the fieldname array. Everything else should be pretty intuitive.

Reply:
In the end I agree, I don't think it's worth asking for such an overhead _just_ to be able to throw those two errors. It will be less intuitive but still fairly obvious why ChoiceScript is falling over, I think.